### ISSUE TITLE
CA1062: False positive with C# 9.0 "or" pattern combinator

### ISSUE THREAD

**Issue by JeremyCaney** [dotnet/roslyn-analyzers#4496]
### Analyzer

**Diagnostic ID**: [CA1062](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1062): "Validate arguments of public methods"

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

**CA1062** throws a false positive if a parameter has been validated using C# 9.0's new `or` pattern combinator as part of its improved pattern matching syntax.

### Steps To Reproduce

```cs
public void DoSomething(Object parameter) 
{
  if (parameter is null or Exception) 
  {
    return;
  }
  _ = parameter.ToString();
}
```

### Expected behavior

If a previous guard clause evaluates a parameter as `is null`, subsequent references to members of that parameter—even if the guard clause uses an `or` pattern combinator as part of its validation logic.

### Actual behavior

> In externally visible method 'void MyClass.DoSomething(object parameter)', validate parameter 'parameter' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.

## Additional context

I haven't extensively tested other pattern matching scenarios. There may be further false positives related to this which are not outlined above.

**Comment by Youssef1313** [comment]
Sounds like a good opportunity to explore the data flow area :smile:

**Comment by Youssef1313** [comment]
Related to https://github.com/dotnet/roslyn-analyzers/pull/4060

@mavasani, Since https://github.com/dotnet/roslyn/issues/47068 won't be implemented, we would need to handle binary operations here. It sounds more complex than I initially thought. Any guidance?
